File: app/src/main/java/com/example/ikastyle/UI/GearDialogFragment.java
package com.example.ikastyle.UI;

import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.os.AsyncTask;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.fragment.app.DialogFragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;

import com.example.ikastyle.Common.Const.GearKind;
import com.example.ikastyle.Common.Util;
import com.example.ikastyle.Dao.ClothingGearDao;
import com.example.ikastyle.Dao.HeadGearDao;
import com.example.ikastyle.Dao.ShoesGearDao;
import com.example.ikastyle.Database.AppDatabase;
import com.example.ikastyle.R;

import java.util.List;

/*
 * ギア一覧を表示するダイアログ
 */
public class GearDialogFragment extends DialogFragment {
    private RecyclerView recyclerView;
    private final GearKind gearKind;

    public GearDialogFragment(GearKind gearKind){
        this.gearKind = gearKind;
    }

    // このinterfaceを呼び出し元フラグメントで実装することで呼び出し元フラグメントの更新を行う
    public interface GearDialogListener{
        void onListItemClick(GearDialogFragment dialogFragment, GearKind gearKind, int gearId);
    }

    private GearDialogListener listener;

    @Override
    public void onAttach(@NonNull Context context){
        super.onAttach(context);

        if(getParentFragment() == null){
            throw new NullPointerException("GearDialogFragmentの親フラグメントが見つかりません");
        }else if(!(getParentFragment() instanceof  GearDialogListener)){
            throw new ClassCastException(getParentFragment() + "はGearDialogListenerを実装していません");
        } else {
            listener = (GearDialogListener) getParentFragment();
        }
    }

    @Override
    @NonNull
    public Dialog onCreateDialog(Bundle savedInstanceState){
        AlertDialog.Builder builder = new AlertDialog.Builder(getActivity());

        // fragment_gear_dialog.xmlのレイアウトをViewとしてインスタンス化
        LayoutInflater inflater = (LayoutInflater) getActivity().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        View content = inflater.inflate(R.layout.fragment_gear_dialog, null);

        // fragment_gear_dialog.xml上のRecyclerViewを取得しレイアウト設定
        recyclerView = content.findViewById(R.id.recyclerView_gear);
        recyclerView.setHasFixedSize(true);
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(content.getContext());
        recyclerView.setLayoutManager(layoutManager);

        // DBからギアデータのリストを取得しRecyclerViewに表示
        AppDatabase db = AppDatabase.getDatabase(getContext());
        GetGearListAsyncTask task = new GetGearListAsyncTask(db, gearKind);
        task.execute();

        // ダイアログにキャンセルボタン設置
        builder.setNegativeButton(getString(R.string.buttonNavigation_cancel), null);

        // ダイアログにViewを設定
        builder.setView(content);

        return builder.create();
    }

    /*
     * ギアリストをDBから非同期で取得
     */
    private class GetGearListAsyncTask extends AsyncTask<Void, Void, Integer> {
        private final AppDatabase db;
        private final GearKind gearKind;
        private final int languageCode = Util.getLanguageCode();

        private List<?> gearList;


        public GetGearListAsyncTask(AppDatabase db, GearKind gearKind) {
            this.db = db;
            this.gearKind = gearKind;
        }

        @Override
        protected Integer doInBackground(Void... params) {
            //実際にDBにアクセスし結果を取得
            switch(gearKind){
                case HEAD:
                    HeadGearDao headGearDao = db.headGearDao();
                    gearList = headGearDao.getHeadGearList(languageCode);
                    break;
                case CLOTHING:
                    ClothingGearDao clothingGearDao = db.clothingGearDao();
                    gearList = clothingGearDao.getClothingGearList(languageCode);
                    break;
                case SHOES:
                    ShoesGearDao shoesGearDao = db.shoesGearDao();
                    gearList = shoesGearDao.getShoesGearList(languageCode);
                    break;
            }

            return 0;
        }

        @Override
        protected void onPostExecute(Integer code){
            GearRecyclerViewAdapter adapter = new GearRecyclerViewAdapter(gearList);
            recyclerView.setAdapter(adapter);
            // ギアリストのItemを選択した時の処理
            adapter.setOnItemClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    int gearId = ((GearListItemLinearLayout)view).getGearId();
                    GearKind gearKind = ((GearListItemLinearLayout)view).getGearKind();
                    // 元Fragment(NewFragment)にコールバック
                    listener.onListItemClick(GearDialogFragment.this, gearKind, gearId);
                    // ダイアログを消す
                    dismiss();
                }
            });
        }
    }
}

File: app/src/main/java/com/splatool/ikastyle/ui/GearDialogFragment.kt
package com.splatool.ikastyle.ui

import android.app.AlertDialog
import android.app.Dialog
import android.content.Context
import com.splatool.ikastyle.common.const.GearKind
import com.splatool.ikastyle.R
import androidx.recyclerview.widget.RecyclerView
import android.os.Bundle
import android.view.LayoutInflater
import androidx.recyclerview.widget.LinearLayoutManager
import com.splatool.ikastyle.model.data.database.AppDatabase
import android.os.AsyncTask
import android.view.View
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.splatool.ikastyle.common.Util
import com.splatool.ikastyle.databinding.FragmentGearDialogBinding
import com.splatool.ikastyle.model.data.entity.ClothingGear
import com.splatool.ikastyle.model.data.entity.HeadGear
import com.splatool.ikastyle.model.data.entity.Loadout
import com.splatool.ikastyle.model.data.entity.ShoesGear
import com.splatool.ikastyle.model.data.repository.GearRepository
import com.splatool.ikastyle.model.data.repository.MainCategoryRepository
import com.splatool.ikastyle.viewModel.GearDialogViewModel
import com.splatool.ikastyle.viewModel.NewViewModel
import com.splatool.ikastyle.viewModel.StoreViewModel
import java.lang.ClassCastException
import java.lang.NullPointerException
import java.util.ArrayList

/*
 * ギア一覧を表示するダイアログ
 */
class GearDialogFragment(private val gearKind: GearKind) : DialogFragment() {
    lateinit var gearAdapter: GearRecyclerViewAdapter
    lateinit var binding : FragmentGearDialogBinding

    // このinterfaceを呼び出し元フラグメントで実装することで呼び出し元フラグメントの更新を行う
    interface GearDialogListener {
        fun onListItemClick(
            dialogFragment: GearDialogFragment,
            gearKind: GearKind,
            gearId: Int
        )
    }

    private lateinit var listener: GearDialogListener
    override fun onAttach(context: Context) {
        super.onAttach(context)
        listener = when (parentFragment) {
            null -> {
                throw NullPointerException("GearDialogFragmentの親フラグメントが見つかりません")
            }
            !is GearDialogListener -> {
                throw ClassCastException(parentFragment.toString() + "はGearDialogListenerを実装していません")
            }
            else -> {
                parentFragment as GearDialogListener
            }
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val builder = AlertDialog.Builder(activity)

        val db : AppDatabase = AppDatabase.getDatabase(requireContext())
        val gearRepository = GearRepository(db.headGearDao(), db.clothingGearDao(), db.shoesGearDao())

        val gearViewModel = ViewModelProvider(this, GearDialogViewModel.GearFactory(gearRepository, gearKind))[GearDialogViewModel::class.java]

        // fragment_gear_dialog.xmlのレイアウトをViewとしてインスタンス化
        val inflater = activity?.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
        val content = inflater.inflate(R.layout.fragment_gear_dialog, null)

        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_gear_dialog, null, false)

        gearAdapter = GearRecyclerViewAdapter(gearViewModel)

        val manager: RecyclerView.LayoutManager = LinearLayoutManager(content.context)
        binding.recyclerViewGear.apply {
            setHasFixedSize(true)
            layoutManager = manager
            adapter = gearAdapter
        }

        // ダイアログにキャンセルボタン設置
        builder.setNegativeButton(getString(R.string.buttonNavigation_cancel), null)

        // ダイアログにViewを設定
        builder.setView(binding.root)

        observeViewModel(gearViewModel)
        return builder.create()
    }

    private fun observeViewModel(viewModel: GearDialogViewModel){
        val gearListObserver = Observer<ArrayList<*>> {
            it.let{
                gearAdapter.setGearList(it)
            }
        }

        val selectedObserver = Observer<Int?>{
            it.let{
                listener.onListItemClick(this@GearDialogFragment, gearKind, it)
                dismiss()
            }
        }

        viewModel.getGearListLiveData().observe(this, gearListObserver)
        viewModel.getSelectedGearIdLiveData().observe(this, selectedObserver)
    }
}